docs: Add GitHub basic setup where best practices are replaced with links

mvincerx opened the pull request:
* Fixes #151
* Create 1-level depth of TOC to simplify
* Add Contributor roles section w links
* Replace longer examples of PR process w links to GitHub workflows
* Revise "Before starting a PR..." to "Best practices.." and simplify

Signed-off-by: Michael Vincerra <michael.vincerra@intel.com>

grahamwhaley commented:
CI is picking up an md link item:

```
INFO[0000] Checking file                                 commit=690110f70cce450dc7376f2b69d010fd06f8f867 file=CONTRIBUTING.md name=kata-check-markdown pid=12317 source=check-markdown version=0.0.1
FATA[0000] file="CONTRIBUTING.md": failed to find heading for link "Contributor-roles" ({Doc:CONTRIBUTING.md Address:Contributor-roles ResolvedPath: Description:Contributor roles Type:internal-link}) - correct link name is "contributor-roles"  commit=690110f70cce450dc7376f2b69d010fd06f8f867 name=kata-check-markdown pid=12317 source=check-markdown version=0.0.1
The command ".ci/static-checks.sh" failed and exited with 1 during .
```

grahamwhaley commented on CONTRIBUTING.md:
```diff
@@ -41,75 +25,97 @@ process documented here.
-## Golang Coding Style
+## Pull requests
+
+All the repositories accept contributions by a GitHub Pull request (PR).
```
I wonder if we want to hotlink out to github for a PR reference? https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-requests

mvincerx replied:
Good call. Thanks.

grahamwhaley commented on CONTRIBUTING.md:
```diff
@@ -41,75 +25,97 @@ process documented here.
+
+- Review [Contributor roles](#Contributor-roles) for special configuration
+
+- [Set up Git](https://help.github.com/en/github/getting-started-with-github/set-up-git). *Note:* The email address you specify must match your primary  GitHub email address.
```
I'm not sure if the 'must match your primary email address' is correct. The email you sign off by must match the email that gets placed into the commit I think, otherwise the CI complains - but, I'm not sure it has to be the primary github email address. Do you know specifics or details @jodh-intel ?

mvincerx replied:
I agree with @grahamwhaley. I'll use "The email address you specify must match the email address that you use to sign-off commits."  This is more accurate.

grahamwhaley commented on CONTRIBUTING.md:
```diff
@@ -41,75 +25,97 @@ process documented here.
+
+- [x] Ensure that [`golangci-lint`](https://github.com/golangci/golangci-lint) is current or has not been installed (the static check scripts will install it if necessary). Changes in either the linters used or the Kata Containers code base can produce spurious errors that do not fail inside the CI systems.
+
+##### Kata runtime
```
@jodh-intel - I think maybe the tests etc. will only work (at least for the runtime repo), if you are in the 'original' repo, and not a personal clone - because of the go path 'fun'... should we check/clarify that these instructions either mention that, or the recommendation due to use of golang is to work in the 'base' clone of the repo (not a personal one), but add a remote to your own personal fork where you push PRs iyswim?

**AH**, I see we cover that below here - maybe we need to ensure what is below (the fork and remote) matches what is above....

jodh-intel commented on CONTRIBUTING.md:
```diff
@@ -687,6 +674,12 @@ This comment is required for the main patch in a sequence. See the following exa
 
+<!---
+  TODO:
+  The 'character limit' below is actually 75. This should be stated. 
```
Agreed. The value is specified here by the way:

- https://github.com/kata-containers/tests/blob/690110f70cce450dc7376f2b69d010fd06f8f867/cmd/checkcommits/checkcommits.go#L58

But we may wish to set the value explicitly in the CI static check script here:

- https://github.com/kata-containers/tests/blob/master/.ci/static-checks.sh#L236..L240

.. by specifying the extra options `--subject-length 75 --body-length 72`.

That would make it easier for this doc to reference if required.

mvincerx replied:
I addressed both of these in new section "Pull request format."

mvincerx commented:
@DougTW I added you to this PR for visibility only.

mvincerx commented:
This PR is still a WIP. See my HTML `<!--notes-->` if questions.

mvincerx commented on CONTRIBUTING.md:
````diff
@@ -41,75 +25,97 @@ process documented here.
+$ popd
+```
+
+**Note:**The final `popd` is required to return to the top-level directory from where other build rules can be executed.
````
Done.  The text starting "The final `popd`..." lacked a preceding empty char. :)

mvincerx commented:
@jodh-intel. 
Please see my latest commit and note following: 
- My primary concern here is: Given there is no `Makefile` in the Community repo, should we scrap this example in favor of a repo that does have one and use a minor change? *IOW: Is this more relevant to user base?*
- Review my new section "fork and clone".  I'm not sure that this method complies with what the Travis CI system anticipates, as far as running a static check. If not, I can revert this back. Nevertheless, it's still not clear how one would run a static check on the original example of `Community` repo if there is no Makefile in that repo--unless the `go get` command is intended to run the static check from a higher level or adjacent repo in the project. Just seems odd. 
- Run a search on my "todo" notes inserted as HTML comments; we can discuss these this week.

jodh-intel commented on CONTRIBUTING.md:
```diff
@@ -361,143 +254,105 @@ $ git log master.. --oneline --decorate=no
-  all of your changes together.
+Since all four commits are related to *the same change*, it makes sense to
+combine all four commits into a *single commit* on your PR. You need to
+[git rebase](https://help.github.com/en/github/using-git/about-git-rebase)
```
i18n point: Although GitHub doesn't appear to offer help in many languages, it *does* offer it in Chinese fwics:

- https://help.github.com/zh/github/using-git/about-git-rebase

Since GitHub appears to auto-redirect based on the browser locale, we might want to consider dropping the hard-coded two-character language code ("en") for all GitHub URLs in this doc to allow the user to get the official help docs in their language (if possible - it seems to default to English). In other words, the URL written above would be:

- https://help.github.com/github/using-git/about-git-rebase

mvincerx replied:
Duly noted. Revised to remove `en` or `zh` going forward.

mvincerx commented on CONTRIBUTING.md:
```diff
@@ -41,313 +27,220 @@ process documented here.
-The other advantage to raising the issue at the outset is that our process requires an
-issue to be associated with every PR (see [patch format](#patch-format)).
+- Review [Contributor roles](#contributor-roles) that require special Git 
+  configuration
```
Done.

mvincerx commented on CONTRIBUTING.md:
```diff
@@ -41,313 +27,220 @@ process documented here.
 
-GitHub has a basic introduction to the PR process
-[here](https://help.github.com/articles/using-pull-requests/).
+* Review [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments) to avoid common `Go` errors.
```
I understand your concern. I wrote "Go Code Review..." becuase that's the [Golang GitHub page's title]( https://github.com/golang/go/wiki/CodeReviewComments); I'm following their convention. I can title it whatever you prefer.

jodh-intel replied:
Sorry - this really is just an "fyi", so feel free to ignore ;)

mvincerx commented on CONTRIBUTING.md:
```diff
@@ -41,313 +27,220 @@ process documented here.
-  > - Creates a *local copy* of the upstream repository and switches to the
-  >   `master` *branch*.
-  > - Creates an "alias" (name) for the upstream repository URL called `origin`.
+* Apply the appropriate GitHub labels to your PR. This
```
Done.

jodh-intel replied:
Great - just seen the note below so let's resolve this one...

mvincerx commented on CONTRIBUTING.md:
```diff
@@ -41,313 +27,220 @@ process documented here.
 
-### Before submitting a PR
+- [Set up Git](https://help.github.com/en/github/getting-started-with-github/set-up-git). 
+  *Note:* The email address you specify must match the email address you use to sign-off commits.
```
Done.

mvincerx commented on CONTRIBUTING.md:
```diff
@@ -687,11 +536,28 @@ This comment is required for the main patch in a sequence. See the following exa
+
+As shown above, pull requests must adhere to these guidelines:
+
+* Preface the PR title with the keyword: ``subsystem:``
```
Thank you for the correction.

mvincerx replied:
Done.

jodh-intel commented on CONTRIBUTING.md:
```diff
@@ -1,38 +1,24 @@
 [Apache License, Version 2.0](https://www.apache.org/licenses/LICENSE-2.0).
 
-It comprises a number of repositories under the [GitHub Kata
+It comprises a number of repositories under the [GitHub* Kata
```
Nit: We tend not to use the asterisk in this context.

mvincerx replied:
Done.

jodh-intel commented on CONTRIBUTING.md:
````diff
@@ -41,313 +27,228 @@ process documented here.
-  $ git clone https://github.com/kata-containers/community
-  $ cd community
-  ```
+  >*Note:* External contributors should use keywords, explained in the link 
````
Nit: For consistency with our other docs, this should strictly be:

```
> **Note:** External ...
```

So that it renders as:

> **Note:** External ...

Same comment for some of the other new "notes" below...

See https://github.com/kata-containers/documentation/blob/master/Documentation-Requirements.md#notes.

jodh-intel commented on CONTRIBUTING.md:
````diff
@@ -41,313 +27,228 @@ process documented here.
+$ cd community
+```
+   
+**Note:** Cloning a forked repository automatically gives a remote `origin`.
````
Strictly, this needs a leading `>` to make the note stand out. Same comment for some further notes below...

> **Note:** Cloning a ...

mvincerx replied:
Done.

jodh-intel commented on CONTRIBUTING.md:
```diff
@@ -41,313 +27,228 @@ process documented here.
-  You need to tell Git that your changes should be uploaded to your fork on
-  the GitHub server when you `git push` This is done by adding what is called
-  a "remote":
+1. Run `git remote -v`. Your remotes should appear similar to these. 
```
Nit: Could you change the final period to a colon?

mvincerx replied:
Done.

mvincerx replied:
There was one last note that needed a formatting correction; done in latest commit.

jodh-intel commented on CONTRIBUTING.md:
````diff
@@ -41,313 +27,228 @@ process documented here.
+    ```
 
-- Make some changes
+    **Warning:** *Never* make changes directly to the `master` branch -
````
> **Warning:** *Never* ...

mvincerx replied:
Done.

jcvenegas commented:
@grahamwhaley @jodh-intel  could you give a new iteration to this PR?

jodh-intel commented on CONTRIBUTING.md:
```diff
@@ -41,313 +27,230 @@ process documented here.
-> The final `popd` is required to return to the top-level directory from where other build rules can be executed.
+- [Set up Git](https://help.github.com/en/github/getting-started-with-github/set-up-git). 
+  
+  >*Note:* The email address you specify must match the email address you 
```
Nit: This should really be `**Note:**`.

jodh-intel commented on CONTRIBUTING.md:
```diff
@@ -41,313 +27,230 @@ process documented here.
+[requirements section of the Kata Developer Guide](https://github.com/kata-containers/documentation/blob/master/Developer-Guide.md#requirements-to-build-individual-components).
 
-  - Your fork (`https://github.com/${your-github-username}/community`) on GitHub
+>*Note*: If you intend to make minor edits, it's acceptable
```
Nit: This should really be `**Note:**`.

mvincerx commented:
fyi: I removed my developer notes and resolved wording changes. I corrected formatting on notes. 
Please take another look. It appears this PR just needs one more approval. 😃

mvincerx commented:
Thank you, @grahamwhaley
